File: src/hotspot/share/utilities/lineReader.cpp
/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 *
 */

#include "precompiled.hpp"
#include "runtime/os.hpp"
#include "utilities/lineReader.hpp"

LineReader::LineReader() {
  _file = nullptr;
  _is_oom = false;
  _buffer = nullptr;
  _buffer_len = 0;
}

LineReader::~LineReader() {
  if (_buffer != nullptr) {
    os::free(_buffer);
  }
}

void LineReader::init(FILE* file) {
  _file = file;
  _buffer_len = 16; // start at small size to test expansion logic
  _buffer = (char*)os::malloc(_buffer_len, mtClass);
  if (_buffer == nullptr) {
    _is_oom = true;
  }
}

char* LineReader::read_line() {
  if (_is_oom) {
    return nullptr;
  }

  int line_len = 0; // the number of characters we have read so far (excluding the trailing \0)
  while (true) {
    assert(line_len < _buffer_len, "sanity");
    int free_space = _buffer_len - line_len;
    char* p = _buffer + line_len;
    int new_len = 0;
    if (fgets(p, free_space, _file) == nullptr) {
      // _file is at EOF
      if (line_len == 0) {
        return nullptr; // EOF
      } else {
        // We have read something in previous loop iteration(s). Return that.
        // The next call to read_line() will return nullptr to indicate EOF.
        return _buffer;
      }
    }

    // fgets() reads at most free_space characters, including the trailing \0, so
    // strlen(p) must be smaller than INT_MAX, and can be safely cast to int.
    assert(strlen(p) < INT_MAX, "sanity");
    new_len = (int)strlen(p);

    // _buffer_len will stop at INT_MAX, so we will never be able to read more than
    // INT_MAX chars for a single input line.
    assert(line_len >= 0 && new_len >= 0 && (line_len + new_len) >= 0, "no int overflow");

    line_len += new_len; // We have read line_len chars so far.

    assert(line_len < _buffer_len, "sanity");
    assert(_buffer[line_len] == '\0', "sanity");

    if (_buffer[line_len - 1] == '\n' || feof(_file)) {
      // We have read an entire line, or reached EOF
      return _buffer;
    }

    if (line_len == _buffer_len - 1) {
      // The buffer is not big enough to hold the entire input line. Expand it.
      if (_buffer_len == INT_MAX) {
        _is_oom = true; // cannot expand anymore.
        return nullptr;
      }
      int new_len = _buffer_len * 2;
      if (new_len < _buffer_len) { // overflows int
        new_len = INT_MAX;
      }
      assert(new_len > _buffer_len, "must be");

      char* new_buffer = (char*)os::realloc(_buffer, new_len, mtClass);
      if (new_buffer == nullptr) {
        _is_oom = true; // oom
        return nullptr;
      } else {
        _buffer = new_buffer;
        _buffer_len = new_len;
      }
    }
  }
}

File: src/jdk.internal.md/share/classes/jdk/internal/org/commonmark/internal/util/LineReader.java
/*
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

/*
 * This file is available under and governed by the GNU General Public
 * License version 2 only, as published by the Free Software Foundation.
 * However, a notice that is now available elsewhere in this distribution
 * accompanied the original version of this file, and, per its terms,
 * should not be removed.
 */

package jdk.internal.org.commonmark.internal.util;

import java.io.Closeable;
import java.io.IOException;
import java.io.Reader;

/**
 * Reads lines from a reader like {@link java.io.BufferedReader} but also returns the line terminators.
 * <p>
 * Line terminators can be either a line feed {@code "\n"}, carriage return {@code "\r"}, or a carriage return followed
 * by a line feed {@code "\r\n"}. Call {@link #getLineTerminator()} after {@link #readLine()} to obtain the
 * corresponding line terminator. If a stream has a line at the end without a terminator, {@link #getLineTerminator()}
 * returns {@code null}.
 */
public class LineReader implements Closeable {

    // Same as java.io.BufferedReader
    static final int CHAR_BUFFER_SIZE = 8192;
    static final int EXPECTED_LINE_LENGTH = 80;

    private Reader reader;
    private char[] cbuf;

    private int position = 0;
    private int limit = 0;

    private String lineTerminator = null;

    public LineReader(Reader reader) {
        this.reader = reader;
        this.cbuf = new char[CHAR_BUFFER_SIZE];
    }

    /**
     * Read a line of text.
     *
     * @return the line, or {@code null} when the end of the stream has been reached and no more lines can be read
     */
    public String readLine() throws IOException {
        StringBuilder sb = null;
        boolean cr = false;

        while (true) {
            if (position >= limit) {
                fill();
            }

            if (cr) {
                // We saw a CR before, check if we have CR LF or just CR.
                if (position < limit && cbuf[position] == '\n') {
                    position++;
                    return line(sb.toString(), "\r\n");
                } else {
                    return line(sb.toString(), "\r");
                }
            }

            if (position >= limit) {
                // End of stream, return either the last line without terminator or null for end.
                return line(sb != null ? sb.toString() : null, null);
            }

            int start = position;
            int i = position;
            for (; i < limit; i++) {
                char c = cbuf[i];
                if (c == '\n') {
                    position = i + 1;
                    return line(finish(sb, start, i), "\n");
                } else if (c == '\r') {
                    if (i + 1 < limit) {
                        // We know what the next character is, so we can check now whether we have
                        // a CR LF or just a CR and return.
                        if (cbuf[i + 1] == '\n') {
                            position = i + 2;
                            return line(finish(sb, start, i), "\r\n");
                        } else {
                            position = i + 1;
                            return line(finish(sb, start, i), "\r");
                        }
                    } else {
                        // We don't know what the next character is yet, check on next iteration.
                        cr = true;
                        position = i + 1;
                        break;
                    }
                }
            }

            if (position < i) {
                position = i;
            }

            // Haven't found a finished line yet, copy the data from the buffer so that we can fill
            // the buffer again.
            if (sb == null) {
                sb = new StringBuilder(EXPECTED_LINE_LENGTH);
            }
            sb.append(cbuf, start, i - start);
        }
    }

    /**
     * Return the line terminator of the last read line from {@link #readLine()}.
     *
     * @return {@code "\n"}, {@code "\r"}, {@code "\r\n"}, or {@code null}
     */
    public String getLineTerminator() {
        return lineTerminator;
    }

    @Override
    public void close() throws IOException {
        if (reader == null) {
            return;
        }
        try {
            reader.close();
        } finally {
            reader = null;
            cbuf = null;
        }
    }

    private void fill() throws IOException {
        int read;
        do {
            read = reader.read(cbuf, 0, cbuf.length);
        } while (read == 0);
        if (read > 0) {
            limit = read;
            position = 0;
        }
    }

    private String line(String line, String lineTerminator) {
        this.lineTerminator = lineTerminator;
        return line;
    }

    private String finish(StringBuilder sb, int start, int end) {
        int len = end - start;
        if (sb == null) {
            return new String(cbuf, start, len);
        } else {
            return sb.append(cbuf, start, len).toString();
        }
    }
}
